File: src/main/java/com/gregtechceu/gtceu/api/sync_system/codecs/RecipeChanceCachesCodec.java
package com.gregtechceu.gtceu.api.sync_system.codecs;

import com.gregtechceu.gtceu.api.capability.recipe.RecipeCapability;
import com.gregtechceu.gtceu.api.registry.GTRegistries;
import com.gregtechceu.gtceu.api.sync_system.ContextualFieldCodec;

import net.minecraft.resources.ResourceLocation;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import it.unimi.dsi.fastutil.objects.Object2IntMap;
import org.jetbrains.annotations.Nullable;

import java.util.IdentityHashMap;

public final class RecipeChanceCachesCodec
                                           implements
                                           ContextualFieldCodec<IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>>> {

    public static final RecipeChanceCachesCodec INSTANCE = new RecipeChanceCachesCodec();

    private RecipeChanceCachesCodec() {}

    @Override
    public JsonElement serializeField(IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>> value,
                                      Context<IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>>> context) {
        JsonObject chanceCache = new JsonObject();
        IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>> currentValue = context.currentValue();
        if (currentValue == null) {
            return chanceCache;
        }

        for (var entry : currentValue.entrySet()) {
            RecipeCapability<?> capability = entry.getKey();
            JsonArray cacheJson = new JsonArray();
            for (Object2IntMap.Entry<?> cacheEntry : entry.getValue().object2IntEntrySet()) {
                JsonObject entryJson = new JsonObject();
                writeEntryJson(entryJson, capability, cacheEntry.getKey(), cacheEntry.getIntValue(), context);
                cacheJson.add(entryJson);
            }
            chanceCache.add(GTRegistries.RECIPE_CAPABILITIES.getKey(capability).toString(), cacheJson);
        }

        return chanceCache;
    }

    @Override
    public @Nullable IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>> deserializeField(
                                                                                             JsonElement value,
                                                                                             Context<IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>>> context) {
        if (!value.isJsonObject()) {
            return context.currentValue();
        }
        IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>> currentValue = context.currentValue();
        if (currentValue == null) {
            return null;
        }

        JsonObject json = value.getAsJsonObject();
        for (var entry : json.entrySet()) {
            RecipeCapability<?> capability = GTRegistries.RECIPE_CAPABILITIES
                    .get(ResourceLocation.parse(entry.getKey()));
            if (capability == null || !entry.getValue().isJsonArray()) {
                continue;
            }
            Object2IntMap<Object> map = getOrCreateCache(currentValue, capability);
            for (JsonElement entryJson : entry.getValue().getAsJsonArray()) {
                if (!entryJson.isJsonObject()) {
                    continue;
                }
                JsonObject entryObject = entryJson.getAsJsonObject();
                Object cacheKey = readEntryJson(capability, entryObject, context);
                map.put(cacheKey, entryObject.get("cached_chance").getAsInt());
            }
        }
        return currentValue;
    }

    @SuppressWarnings("unchecked")
    private static Object2IntMap<Object> getOrCreateCache(
                                                          IdentityHashMap<RecipeCapability<?>, Object2IntMap<?>> currentValue,
                                                          RecipeCapability<?> capability) {
        return (Object2IntMap<Object>) currentValue.computeIfAbsent(capability,
                key -> ((RecipeCapability<Object>) key).makeChanceCache());
    }

    @SuppressWarnings("unchecked")
    private static void writeEntryJson(JsonObject json,
                                       RecipeCapability<?> capability,
                                       @Nullable Object content,
                                       int chance,
                                       Context<?> context) {
        RecipeCapability<Object> typedCapability = (RecipeCapability<Object>) capability;
        json.add("entry", typedCapability.serializer.toJson(typedCapability.of(content), context.lookup()));
        json.addProperty("cached_chance", chance);
    }

    @SuppressWarnings("unchecked")
    private static Object readEntryJson(RecipeCapability<?> capability, JsonObject json, Context<?> context) {
        return ((RecipeCapability<Object>) capability).serializer.fromJson(json.get("entry"), context.lookup());
    }
}

File: src/main/java/com/gregtechceu/gtceu/api/sync_system/codecs/RecipeChanceCachesCodec.kt
package com.gregtechceu.gtceu.api.sync_system.codecs

import com.gregtechceu.gtceu.api.capability.recipe.RecipeCapability
import com.gregtechceu.gtceu.api.registry.GTRegistries
import com.gregtechceu.gtceu.api.sync_system.ContextualFieldCodec

import net.minecraft.resources.ResourceLocation

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import it.unimi.dsi.fastutil.objects.Object2IntMap

import java.util.IdentityHashMap

class RecipeChanceCachesCodec private constructor() : ContextualFieldCodec<IdentityHashMap<RecipeCapability<*>, Object2IntMap<*>>> {

	override fun serializeField(
		value: IdentityHashMap<RecipeCapability<*>, Object2IntMap<*>>,
		context: ContextualFieldCodec.Context<IdentityHashMap<RecipeCapability<*>, Object2IntMap<*>>>,
	): JsonElement {
		val chanceCache = JsonObject()
		for ((capability, cache) in value) {
			val cacheJson = JsonArray()
			for (cacheEntry in cache.object2IntEntrySet()) {
				val entryJson = JsonObject()
				writeEntryJson(entryJson, capability, cacheEntry.key, cacheEntry.intValue, context)
				cacheJson.add(entryJson)
			}
			chanceCache.add(GTRegistries.RECIPE_CAPABILITIES.getKey(capability)!!.toString(), cacheJson)
		}

		return chanceCache
	}

	override fun deserializeField(
		value: JsonElement,
		context: ContextualFieldCodec.Context<IdentityHashMap<RecipeCapability<*>, Object2IntMap<*>>>,
	): IdentityHashMap<RecipeCapability<*>, Object2IntMap<*>>? {
		if (!value.isJsonObject) return context.currentValue
		val currentValue = context.currentValue ?: return null

		for ((key, cacheJson) in value.asJsonObject.entrySet()) {
			val capability = GTRegistries.RECIPE_CAPABILITIES.get(ResourceLocation.parse(key))
			if (capability == null || !cacheJson.isJsonArray) continue

			val map = getOrCreateCache(currentValue, capability)
			for (entryJson in cacheJson.asJsonArray) {
				if (!entryJson.isJsonObject) continue
				val entryObject = entryJson.asJsonObject
				val cacheKey = readEntryJson(capability, entryObject, context)
				map.put(cacheKey, entryObject.get("cached_chance").asInt)
			}
		}
		return currentValue
	}

	@Suppress("UNCHECKED_CAST")
	private fun getOrCreateCache(currentValue: IdentityHashMap<RecipeCapability<*>, Object2IntMap<*>>, capability: RecipeCapability<*>): Object2IntMap<Any?> =
		currentValue.computeIfAbsent(capability) { key ->
			(key as RecipeCapability<Any?>).makeChanceCache()
		} as Object2IntMap<Any?>

	@Suppress("UNCHECKED_CAST")
	private fun writeEntryJson(json: JsonObject, capability: RecipeCapability<*>, content: Any?, chance: Int, context: ContextualFieldCodec.Context<*>) {
		val typedCapability = capability as RecipeCapability<Any?>
		json.add("entry", typedCapability.serializer.toJson(typedCapability.of(content), context.lookup))
		json.addProperty("cached_chance", chance)
	}

	@Suppress("UNCHECKED_CAST")
	private fun readEntryJson(capability: RecipeCapability<*>, json: JsonObject, context: ContextualFieldCodec.Context<*>): Any? =
		(capability as RecipeCapability<Any?>).serializer.fromJson(json.get("entry"), context.lookup)

	companion object {
		@JvmField
		val INSTANCE: RecipeChanceCachesCodec = RecipeChanceCachesCodec()
	}
}
